### PR TITLE
[iam/sso] refactor client config middleware

### PR DESCRIPTION
**Pull request by AlexTugarev**
This addresses https://github.com/gitpod-io/gitpod/pull/15440#discussion_r1053534414

... to separate error handling in the two separate cases – `/start` and `/callback`. The different expectations are covered in tests.

## Related Issue(s)
Follow up to #15440

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/iam/pkg/oidc/service.go [hunk 1]
```diff
@@ -157,30 +157,34 @@ func randString(size int) (string, error) {
...
+	}
+
+	state, err := decodeStateParam(stateParam)
+	if err != nil {
+		return nil, fmt.Errorf("bad state param")
+	}
+	config := s.configsById[state.ClientConfigID]
+	if config != nil {
+		return config, nil
 	}
```
Comment (easyCZ) [inline components/iam/pkg/oidc/service.go:188]: Why do we need to modify the service state inside a middleware? This causes the middleware to be coupled to the service.

If we need to extract config, we can just call a function to extract it as part of the handler, and not store it. This feels like a premature optimization which makes it hard to follow what is happening as the middleware causes side-effects.

Comment (easyCZ) [reply components/iam/pkg/oidc/service.go:188]: It would look like this https://github.com/gitpod-io/gitpod/compare/at/refactor-config-mw-2?expand=1

Comment (AlexTugarev) [reply components/iam/pkg/oidc/service.go:188]: I do not agree on that at all, as it creates bloated handlers, and the argument sounds like it's just a personal taste. 
Anyways, for the sake of getting things done, I'll rewrite just to not loose focus on the more important things.

Comment (AlexTugarev) [reply components/iam/pkg/oidc/service.go:188]: done

Comment (easyCZ) [reply components/iam/pkg/oidc/service.go:188]: >, and the argument sounds like it's just a personal taste

It's not personal taste. It's based on the following.
* Explicit is better than implicit. If I can make something very direct, without indirection, it's often easier to read, and hence get it correct, than indirect (`provider` is the indirection here)
* In this example, we're not re-using that middleware for multiple routes. As such, we get better readability by handling it directly, rather than through an indirection - middleware
* I also want to reduce bloat. And for me, bloat are the things that are there, which do not need to be there - concepts, indirections, abstractions. I mentioned this, because for me this felt like an unnecessary indirection.

Thanks for making the change.
